### DISCUSSION TITLE
MacOS game controller not detected

### DISCUSSION THREAD

**Discussion by palmerj** [mamedev/discussions#167] [General]
I'm running mame 0.272 from macos [homebrew](https://formulae.brew.sh/formula/mame) on MacOS 15.2 (24C101) and I can't get my gamepads detected. I'm testing with both a bluetooth PS5 controller and a 8bitdo M30. 

The game controllers are detected with the [SDL2 gamepad tester ](https://generalarcade.com/gamepadtool/)tool and shows up in the MacOS system settings.

I've tried setting joystickprovider to auto, sdlgame, and sdljoy and still not luck.

Verbose output from launching mame with `mame mk2 -window -joystick -verbose` is here: https://gist.github.com/palmerj/f87f19c6b3dfe479f633397d4fa67c23

mame.ini Config is here: https://gist.github.com/palmerj/689c521192e06458c90340bc9b0d913c

Any help would be greatly appreciated. 

<img width="717" alt="Screenshot 2024-12-17 at 07 00 54" src="https://github.com/user-attachments/assets/4ac79093-fdbf-4727-8911-2b75133351bf" />

<img width="571" alt="Screenshot 2024-12-17 at 07 01 40" src="https://github.com/user-attachments/assets/ea6c5bc2-9655-41de-a5fa-241332b21fc2" />

**Comment by palmerj** [comment]
Can anyone help?

**Reply by rb6502** [reply]
Controller detection is completely a function of SDL.  Your verbose output shows SDL did not see any controllers for whatever reason.

**Reply by rb6502** [reply]
Try -joystickprovider sdljoy since the higher-level SDL controller functions don't seem to be interested in cooperating.

**Comment by palmerj** [comment]
So more to this. I added some debugging when having both my PS5 and 8bitdo M30 (xbox mode) connected.  I added some debugging code to the SDL2 gamepad module:

```
diff --git a/src/osd/modules/input/input_sdl.cpp b/src/osd/modules/input/input_sdl.cpp
index f1fb0241b40..65d7f04d26c 100644
--- a/src/osd/modules/input/input_sdl.cpp
+++ b/src/osd/modules/input/input_sdl.cpp
@@ -2677,9 +2677,10 @@ public:
 
 		sdl_joystick_module_base::input_init(machine);
 
-		osd_printf_verbose("Game Controller: Start initialization\n");
+		osd_printf_verbose("Game Controller: Start initialization num:%d\n", SDL_NumJoysticks());
 		for (int physical_stick = 0; physical_stick < SDL_NumJoysticks(); physical_stick++)
 		{
+			osd_printf_verbose("Game Controller: Start initialization physical_stick num:%d, IsGameController: %d\n", physical_stick, SDL_IsGameController(physical_stick));
 			// try to open as a game controller
 			SDL_GameController *ctrl = nullptr;
 			if (m_initialized_game_controller && SDL_IsGameController(physical_stick))
```

I've even added mapping for my 2 controllers into a custom db txt file:

```
030000005e0400008e02000014010000,6 Button,platform:Mac OS X,a:b0,b:b1,x:b2,y:b3,back:b9,guide:b8,start:b10,leftshoulder:b4,rightshoulder:a2,dpup:b11,dpdown:b12,dpleft:b13,dpright:b14,lefttrigger:b5,righttrigger:a5,
030000004c050000e60c000000010000,PS5 Controller,a:b1,b:b2,back:b8,dpdown:h0.4,dpleft:h0.8,dpright:h0.2,dpup:h0.1,guide:b12,leftshoulder:b4,leftstick:b10,lefttrigger:a3,leftx:a0,lefty:a1,misc1:b14,rightshoulder:b5,rightstick:b11,righttrigger:a4,rightx:a2,righty:a5,start:b9,touchpad:b13,x:b0,y:b3,platform:Mac OS X,
```

Running `./mame -controller_map gamecontrollerdb.txt -window mk2 -verbose` I get the following:

```
Game Controller: 2 controller mapping(s) added from file [gamecontrollerdb.txt].
Game Controller: Start initialization num:0
Game Controller: End initialization
```

So SDL_NumJoysticks() is returning 0. 

I've also built against a custom build version of SDL2 ensuring that MFI is built in:

```
mkdir build
cd build
../configure --enable-joystick-mfi
 make -j10
 make install
 ```

And tried the official downloaded framework for SDL2-2.30.10. Still no luck....

@rb6502 Sorry to bother but I see you wrote much of this SDL2 input code. Any idea?

**Comment by palmerj** [comment]
OK definitely something wrong with something the way SDL2 is looking for gamecontroller. This test code doesn't work:

```c
#include <SDL.h>
#include <stdio.h>

int main(int argc, char* argv[]) {
    // Initialize SDL with joystick subsystem
    if (SDL_Init(SDL_INIT_JOYSTICK) < 0) {
        printf("SDL initialization failed: %s\n", SDL_GetError());
        return 1;
    }

    SDL_JoystickEventState(SDL_ENABLE);

    int num_joysticks = SDL_NumJoysticks();
    printf("Number of joysticks detected: %d\n", num_joysticks);

    for (int i = 0; i < num_joysticks; i++) {
        SDL_Joystick* joystick = SDL_JoystickOpen(i);
        if (joystick) {
            printf("\nJoystick %d:\n", i);
            printf("Name: %s\n", SDL_JoystickName(joystick));
            printf("Number of axes: %d\n", SDL_JoystickNumAxes(joystick));
            printf("Number of buttons: %d\n", SDL_JoystickNumButtons(joystick));
            printf("Number of hats: %d\n", SDL_JoystickNumHats(joystick));
            SDL_JoystickClose(joystick);
        } else {
            printf("Failed to open joystick %d: %s\n", i, SDL_GetError());
        }
    }

    printf("\nPress Enter to exit...");
    getchar();

    SDL_Quit();

    return 0;
}
```

`gcc joy_test.c -o joy_test $(sdl2-config --cflags --libs)`

```
./joy_test
Number of joysticks detected: 0

Press Enter to exit...
```

**Comment by palmerj** [comment]
OK i think this might be a regression caused by https://github.com/libsdl-org/SDL/commit/63aff8e3d0a7d9cbc0e306051193a1b23c229967 which was released in SDL2-2.30.9+. SDL2-2.30.8 works ok and the xbox controller is found. So I need to raised an issue with the SDL project

The PS5 controller issue was a separate issue can was fixed with a repairing.

**Reply by rb6502** [reply]
Nice work tracking that down!  That explains why it works for me, I haven't updated since 2.30.1.

**Reply by palmerj** [reply]
I've reported the bug in the SDL project here https://github.com/libsdl-org/SDL/issues/11742. Would the MAME project be open to a patch to workaround this issue for the mean time?

**Reply by palmerj** [reply]
Ok seems like the SLD devs are saying this is a feature of SLD. So MAME would need some sort of delay or polling of devices...

**Reply by palmerj** [reply]
I could add a hack which adds a SDL_Delay (1000ms seems to be needed) like so:

```
 2674  +             // Give system time to detect initial controllers                                              
 2675  +             Uint32 start = SDL_GetTicks();                                                                 
 2676  +             while ((SDL_GetTicks() - start) < 1000) {  // 100ms timeout                                    
 2677  +                 SDL_PumpEvents();  // Process events without removing them from queue                      
 2678  +                 SDL_Delay(1);                                                                              
 2679  +             }        
 ```
 
or maybe rather than adding controllers during the module input_init start it should be picked up as a device event and added when the SDL event is processed - but I don't know the MAME architecture very well, and maybe that won't work...

**Reply by cuavas** [reply]
MAME needs to know what game controllers are available on start to set up default assignments.

Adding a delay of one second before getting anything on the screen definitely isn’t a good look.

**Reply by rb6502** [reply]
Would SDL_PollEvent() work?  I'd rather not #ifdef the Core Foundation call, but it's still a better solution than an enforced 1 second delay.

**Reply by palmerj** [reply]
> MAME needs to know what game controllers are available on start to set up default assignments.

OK thank you. 

> Would SDL_PollEvent() work? I'd rather not #ifdef the Core Foundation call, but it's still a better solution than an enforced 1-second delay.

SDL_PollEvent would not be ideal as it will process other SDL events we don't want in the controller input module (e.g window, mouse, or key). But something like what I've shown above using SDL_PumpEvents is OK.

Something like the following could be done, but it would only create a 100ms delay. But I do understand the issue including a non-sdl lib in MAME as a direct dependency, even though it should be available on all MacOS systems by default:

```c++
#ifdef __APPLE__
#include <CoreFoundation/CoreFoundation.h>
#endif


#ifdef __APPLE__
       // Check SDL version
       SDL_version ver;
       SDL_GetVersion(&ver);
       if (ver.major == 2 && ((ver.minor == 30 && ver.patch >= 9) || ver.minor > 30)) {
           // Process IOKit events to ensure controller enumeration on SDL >= 2.30.9
           osd_printf_verbose("Game Controller: Processing IOKit events...\n");
           CFRunLoopRunInMode(kCFRunLoopDefaultMode, 0.1, false);
       }
#endif
```

Please let me know what I should do. I can prepare a PR.

**Reply by palmerj** [reply]
I've done a little more testing with my Ps5 and M30 controllers connected. It seems that it generally detects the devices before 400ms (about 390ms to be precise). But 1sec is likely a safer number.  I also did further testing with the CFRunLoopRunInMode approach and it wasn't finding my Ps5 controller at all, even when I bumped it's value up to 1sec... So maybe the SDL_PumpEvents approach is the only robust way in any case.
